crates/air-lib/air-parser/src/parser/validator.rs

```diff
@@ -30,6 +30,194 @@ use multimap::MultiMap;
+                        self.stack
+                            .push((CheckInstructionKind::PivotalNext(left_iterable), span));
+                    }
+                    (Some((CheckInstructionKind::PivotalNext(iterator), _)), Some((..))) => {
+                        // potential failure but need to check when fold pops up.
+                        self.stack.push((CheckInstructionKind::Merging, span));
+                        self.potentially_impossible_spans.push((iterator, span));
+                    }
+                    (Some((..)), Some((CheckInstructionKind::PivotalNext(iterable), _))) => {
+                        self.stack
+                            .push((CheckInstructionKind::PivotalNext(iterable), span));
+                    }
+                    (Some(_), Some(_)) => {
+                        self.stack.push((CheckInstructionKind::Merging, span));
+                    }
+                    _ => {
+                        // disable machine if Merging invariant, namely there must be 2 kinds on a stack, is broken.
+                        self.stack.clear();
+                        self.is_enabled = false;
+                    }
+                }
+            }
+            CheckInstructionKind::ReplacingWithCheck(iterator_name) => {
+                self.replacing_with_check_common(iterator_name, span, &instr_kind);
+            }
+            CheckInstructionKind::PopStack1ReplacingWithCheck(iterator_name) => {
+                self.stack.pop();
+                self.replacing_with_check_common(iterator_name, span, &instr_kind);
+            }
+            CheckInstructionKind::PivotalNext(_) | CheckInstructionKind::Simple => {
+                self.stack.push((instr_kind, span))
+            }
+            CheckInstructionKind::PopStack1 => {
+                self.stack.pop();
+                self.stack.push((instr_kind, span))
+            }
+            CheckInstructionKind::PopStack2 => {
+                self.stack.pop();
+                self.stack.pop();
+                self.stack.push((instr_kind, span))
+            }
+        }
+    }
+
+    fn after_next_check(&mut self, iterable: &'name str) {
+        let error_span = self
+            .potentially_impossible_spans
+            .iter()
+            .find(|(kind_iterable, _)| iterable == *kind_iterable)
+            .map(|(_, span)| span);
+        if let Some(span) = error_span {
+            self.impossible_spans.push(*span);
+        }
+    }
+
+    fn replacing_with_check_common(
+        &mut self,
+        iterator_name: &'name str,
+        span: Span,
+        instr_kind: &CheckInstructionKind<'name>,
+    ) {
+        let child = self.stack.pop();
+        match child {
+            Some((CheckInstructionKind::PivotalNext(checking_iterable), _)) => {
+                self.after_next_check(iterator_name);
+                self.stack
+                    .push((CheckInstructionKind::PivotalNext(checking_iterable), span));
```

Comment:
You are recreating the match. It can be made more explicit if you capture the match and then use it.

Reply:
This forces me to use pattern matching with match guard uses matches! and this doesn't look any better than re-creating a capture.

Reply:
I mean something like this
```
Some((blabla@CheckInstructionKind::PivotalNext(_), _)) => {
  ...
  self.stack.push(blabla);
}
```

Reply:
Fixed